### PR TITLE
Feature/branch coverage

### LINKED ISSUE #380: Support for Path Coverage
Starting with [version 2.3.0](http://xdebug.org/updates.php#x_2_3_0) Xdebug supports [collecting branch coverage data](http://derickrethans.nl/path-branch-coverage.html).

This package and thus phpunit should use this new feature to gather and output branch coverage in addition to the existing sequence/line coverage.

### PR DESCRIPTION
**Pull request by Maks3w**
This PR continue the work @sebastianbergmann made on `feature/path-coverage` branch.

Includes PRs #398 and #399 

Close #380 

TODO:
* Clover XML Report. I won't add support for `cond` line types.

### CODE REVIEW

**Comment by sebastianbergmann** [comment]
Thank you, @Maks3w, for this contribution. It will take a while for me to review this as I am very busy at the moment.

**Comment by Maks3w** [comment]
This PR is finish and ready for review. I suggest start for #398 and #399 for make the changeset more tiny.

**Comment by sebastianbergmann** [comment]
Yes, of course, I will look at #398 and #399 first. In the meantime it would help me (and others) if you could attach screenshots (maybe from the https://github.com/sebastianbergmann/money/ project) to this issue. Thanks!

**Comment by Maks3w** [comment]
These commits show the differences in the output respect master.

https://github.com/Maks3w/php-code-coverage/commit/482cb3fe89a02366c57cb2d173bc0d91f87cd849
https://github.com/Maks3w/php-code-coverage/commit/5e875748e80ceffe3481b3237b0c1e99a83358d3
https://github.com/Maks3w/php-code-coverage/commit/4bc19d0524dbba07505a3505c96dcaa112ca4519
https://github.com/Maks3w/php-code-coverage/commit/91387b83d6bd752ee403c59956c23b732685925d

Basically I've added the Path data close to the line statistics. 


Now I'm playing with executing this in a real library project and I try to make it work. Seems does not recognize the paths but I don't know yet why.

**Comment by sebastianbergmann** [comment]
Ah, okay. So for now it's "only" textual. Still a good starting point. Do you also want to implement something like http://derickrethans.nl/images/content/paths-covered-mockup.png?

**Comment by Maks3w** [comment]
No sorry, too much. I'm trying to use this data for colorize paths in coveralls.io (using coverage 0/1 value instead hits per line)

**Comment by Maks3w** [comment]
I've found a bug related to namespaces. I'll fix it later.

**Comment by Maks3w** [comment]
Now works.

@derickr Any change for add a whitelist coverage feature to XDebug? I think will save lots of memory because it now tracks all the PHPUnit framework stuff.

**Comment by sebastianbergmann** [comment]
@Maks3w http://bugs.xdebug.org/view.php?id=1144 is about pushing the filtering from PHPUnit down to Xdebug.

**Comment by sebastianbergmann** [comment]
Path coverage is always shown as 0% for me:

screenshot from 2015-11-05 15-20-03

**Comment by Maks3w** [comment]
@sebastianbergmann Fixed

**Comment by Maks3w** [comment]
About the HTML Mockup I suggest to do the same I expect to do with coveralls. Turn green or red each line if is covered by a branch or not.

Some tests have too many paths > 4096 which is impossible to print.

**Comment by Maks3w** [comment]
@sebastianbergmann I've restored the compatibility with Xdebug < 2.3.2. However I suggest force pathCoverage param from PHPUnit

**Comment by sebastianbergmann** [comment]
Please rebase against current `master`, thanks.

**Comment by sebastianbergmann** [comment]
BTW: Thank you for your contributions. I am happy that somebody else works on this code :-) I would be even happier if you would configure your Git client to properly store your name and email with your commits (see https://github.com/sebastianbergmann/phpunit/blob/master/CONTRIBUTING.md#workflow).

Keep up the good work!

**Comment by Maks3w** [comment]
Rebased.

I don't understand what do you refer with configure my author data on Git. All my commits have a name and email.

**Comment by Maks3w** [comment]
Something is broken. Tests pass but resume is wrong.

**Comment by sebastianbergmann** [comment]
`Maks3w <github.maks3w@virtualplanets.net>` is what my Git client shows for your commits. I doubt that "Maks3w" is your real name :-) But no worries.

**Comment by Maks3w** [comment]
@sebastianbergmann Done. Was a problem while skipping path coverage if xdebug didn't support

**Comment by sebastianbergmann** [comment]
Re: https://github.com/sebastianbergmann/php-code-coverage/pull/400#issuecomment-154140875

Sure, PHPUnit will need a new configuration option (off by default, IMHO) to control path coverage.

**Comment by Maks3w** [comment]
No, Maks3w it's not my real name and my skin is not yellow  :-)

**Comment by Maks3w** [comment]
I've open https://github.com/sebastianbergmann/phpunit/pull/1937 about path coverage switch on PHPUnit configuration

**Comment by Maks3w** [comment]
Reopen as https://github.com/sebastianbergmann/phpunit/pull/1938

**Comment by sebastianbergmann** [comment]
* The test suite for PHP_CodeCoverage itself fails when path coverage is disabled
* The HTML report must not have path coverage related columns etc. when path coverage is disabled

**Comment by Maks3w** [comment]
I know. I just did the lowest effort so the library still functional. 

Do you think is needed to test both scenarios? (path on/off)

**Comment by Maks3w** [comment]
Exclude the path information in the HTML output will require duplicate the templates or develop a conditional statement the template.

Will also require duplicate all the test fixtures.

I think 0/0 could be a valid value as has been always in the Clover report (`conditionals="0" coveredconditionals="0"`)

Probably the only thing we could do is paint the values (text and html outputs) with a disabled color (gray)

**Comment by Maks3w** [comment]
I've rewritten the bc commit. Loops are enough safe for to not depend of the pathCoverage setting value.

**Comment by Maks3w** [comment]
Tested with 2.4.0beta1

**Comment by derickr** [comment]
How is this coming along?

**Comment by derickr** [comment]
poke?

**Comment by sebastianbergmann** [comment]
Closing #400 and #425 because they have run out-of-sync with `master`. Please coordinate, @Maks3w and @posey2010, and send a new pull request. Thanks!

**Comment by Maks3w** [comment]
@sebastianbergmann I won't spent more time on this for nothing. This PR has been open for 4 months. If you won't merge it why open a new PR

**Comment by sebastianbergmann** [comment]
I did not merge it because there were "competing" / conflicting merge requests and I got confused. I hope that a new pull request will be less confusing. Sorry :-(

**Comment by pgamilde** [comment]
It's 2020, and we badly need this feature..
